### PR TITLE
drivers: dac: Add TI DACx0501 driver

### PR DESCRIPTION
**Pull request by erang20**
Adds a DAC driver for Texas Instruments DACx0501 family of devices

### CODE REVIEW

File: drivers/dac/Kconfig.dacx0501 [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# DAC configuration options
+
+# Copyright (c) 2023 Google, LLC.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config DAC_DACX0501
+	bool "TI DACx0501 DAC driver"
+	default n
```
Comment (bbilas) [inline drivers/dac/Kconfig.dacx0501:9]: That's superfluous.

File: drivers/dac/Kconfig.dacx0501 [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# DAC configuration options
+
+# Copyright (c) 2023 Google, LLC.
```
Comment (bbilas) [inline drivers/dac/Kconfig.dacx0501:3]:
```suggestion
# Copyright (c) 2024 Google, LLC.
```

File: drivers/dac/dac_dacx0501.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+#define DACX0501_MASK_DEVICE_ID_RES	 GENMASK(14, 12)
+#define DACX0501_MASK_CONFIG_REF_PWDWN	 BIT(8)
+#define DACX0501_MASK_CONFIG_DAC_PWDWN	 BIT(0)
+#define DACX0501_MASK_GAIN_BUFF_GAIN	 BIT(0)
+#define DACX0501_MASK_GAIN_REFDIV_EN	 BIT(8)
+#define DACX0501_MASK_TRIGGER_SOFT_RESET (BIT(1) | BIT(3))
+#define DACX0501_MASK_STATUS_REF_ALM	 BIT(0)
+
+/* Specifies the source of the reference voltage. */
+enum RefVoltageSource {
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:41]: Please use snake_case:
```suggestion
enum ref_voltage_source {
```

File: drivers/dac/dac_dacx0501.c [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	REF_INTERNAL, /* Internal 2.5V voltage reference. */
+	REF_EXTERNAL, /* External pin voltage reference. */
+};
+
+/* Specifies the reference voltage multiplier. */
+enum VoltageMult {
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:47]:
```suggestion
enum voltage_mult {
```

File: drivers/dac/dac_dacx0501.c [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	VM_MUL2, /* Multiplies by 2. */
+	VM_NOP,	 /* Multiplies by 1. */
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:49]: I'd suggest to use the following for consistency:
```suggestion
	VM_MUL1,	 /* Multiplies by 1. */
```

File: drivers/dac/dac_dacx0501.c [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	const struct dacx0501_config *config = dev->config;
+	struct dacx0501_data *data = dev->data;
+	uint16_t device_id;
+	int status;
+
+	/* DACx0501 series only has a single output channel. */
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("%s: Unsupported channel %d", __func__, channel_cfg->channel_id);
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:104]: Please remove the `__func__` here and in other locations. The function name is already added by the logging subsystem.

File: drivers/dac/dac_dacx0501.c [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	value <<= (16 - data->resolution);
+
+	return dacx0501_reg_write(dev, DACX0501_REG_DAC, value);
+}
+
+static int dacx0501_init(const struct device *dev)
+{
+	struct dacx0501_data *data = dev->data;
+
+	data->configured = false;
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:179]: Not necessary. False is the default for static boolean variables anyway. You can remove this function and set it to `NULL` below.

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Thanks for the contribution.

Here is a quick review with some first comments.

Please always fix the existing PR with force-pushing and don't open a new one for review updates.

File: dts/bindings/dac/ti,dac-x0501.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    description: |
+      DAC voltage reference select: either internal or external
+
+  voltage_mult:
+    type: string
+    enum:
+      - "mul2"
+      - "nop"
+      - "div2"
+    default: "nop"
```
Comment (decsny) [inline dts/bindings/dac/ti,dac-x0501.yaml:26]: any DT binding default must have a rationale in the description

https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

**Review by decsny — DISMISSED** [review]

File: drivers/dac/dac_dacx0501.c (hunk 3 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c:41]: Done

**Comment by martinjaeger** [comment]
@erang20 Please include the updates in your original commits and leave the comments open so that the reviewer can mark them as resolved. There are some hints in the Zephyr docs how to include review comments using `git rebase -i` and force-pushing.

And please do **not** open another PR, but fix this one. If you need further assistance after reading the docs, just let us know.

File: drivers/dac/Kconfig.dacx0501 (hunk 1 above)
Comment (erang20) [reply drivers/dac/Kconfig.dacx0501:9]: Did you mean to replace it with the auto generated DT_HAS_DAC_DACX0501_ENABLED ?

File: drivers/dac/Kconfig.dacx0501 (hunk 2 above)
Comment (erang20) [reply drivers/dac/Kconfig.dacx0501:3]: Done

File: drivers/dac/dac_dacx0501.c (hunk 6 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c]: Done

File: drivers/dac/dac_dacx0501.c (hunk 5 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c]: Done

File: drivers/dac/dac_dacx0501.c (hunk 4 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c]: Done

File: drivers/dac/dac_dacx0501.c (hunk 7 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c]: Done

File: dts/bindings/dac/ti,dac-x0501.yaml (hunk 8 above)
Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Added

**Comment by erang20** [comment]
Thanks for bearing with review. Hopefully I'm doing it right now :)

**Review by erang20 — COMMENTED** [review]
Thanks for the review!

File: drivers/dac/Kconfig.dacx0501 (hunk 1 above)
Comment (bbilas) [reply drivers/dac/Kconfig.dacx0501:9]: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults

File: dts/bindings/dac/ti,dac-x0501.yaml (hunk 8 above)
Comment (decsny) [reply dts/bindings/dac/ti,dac-x0501.yaml]: it's still not fixed, please look at the link for explanation

Comment (decsny) [reply dts/bindings/dac/ti,dac-x0501.yaml]: In that description in the commit, you don't write the justification which you just wrote in this comment, that's what is required, the reasoning for the default value

Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Added now in the source.

File: dts/bindings/dac/ti,dac-x0501.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2023 Google, LLC.
```
Comment (bbilas) [inline dts/bindings/dac/ti,dac-x0501.yaml:1]: One more ;)

File: drivers/dac/dac_dacx0501.c [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	struct dacx0501_data *data = dev->data;
+	uint16_t device_id;
+	int status;
+
+	/* DACx0501 series only has a single output channel. */
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("Unsupported channel %d", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+	if (data->configured) {
```
Comment (bbilas) [inline drivers/dac/dac_dacx0501.c:107]: Can you add a newline before each if statement?

Comment (erang20) [reply drivers/dac/dac_dacx0501.c:108]: Done

File: dts/bindings/dac/ti,dac-x0501.yaml (hunk 9 above)
Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Done

File: drivers/dac/dac_dacx0501.c [hunk 11]
```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2024 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Driver for Texas Instruments DACx0501 series
+ *
+ * Device driver for the  Texas Instruments DACx0501 series of devices: DAC60501, DAC70501 and
```
Comment (bbilas) [inline drivers/dac/dac_dacx0501.c:10]: I would move this description to a YAML file, but let's hear what others think about it.

Comment (erang20) [reply drivers/dac/dac_dacx0501.c:10]: @martinjaeger Can you address @bbilas comment?

File: dts/bindings/dac/ti,dac-x0501.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    enum:
+      - "mul2"
+      - "mul1"
+      - "div2"
+    default: "mul1"
+    description: |
+      This setting can be used to control the output voltage range within the supported bit
+      resolution. mul2 will double the output range but lower the resolutiuon, while div2 will
+      lower the range but double the resolution. The default is arbitrarily set to the middle
+      point (mul1) which should fit most applications (if they care about it)
```
Comment (decsny) [inline dts/bindings/dac/ti,dac-x0501.yaml:31]: Please change the last sentence to "The default is set to the middle setting mul1 to fit most use cases."

Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Done

File: drivers/dac/dac_dacx0501.c [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	status = dacx0501_reg_write(
+		dev, DACX0501_REG_GAIN,
+		FIELD_PREP(DACX0501_MASK_GAIN_REFDIV_EN, config->voltage_mult == VM_DIV2) |
+			FIELD_PREP(DACX0501_MASK_GAIN_BUFF_GAIN, config->voltage_mult == VM_MUL2));
+	if (status) {
+		LOG_ERR("GAIN Register update failed");
+		return status;
+	}
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:148]: I think this code should all go into the driver init function, similar to the `dacx3608` driver. It is more a device-level check / initialization and would not need to be called for each channel (if we had more than one channel).

File: drivers/dac/dac_dacx0501.c [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+	/* DACx0501 series only has a single output channel. */
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("Unsupported channel %d", channel_cfg->channel_id);
+		return -ENOTSUP;
+	}
+
+	if (data->configured) {
+		LOG_ERR("Channel already configured");
+		return -EINVAL;
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:110]: This return code is not exactly right considering the API docs. As mentioned above, unless the device can really only be configured once, I don't think this is an error.

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Sorry for the delay with my full review. Generally looks good, only got some minor suggestions left for improvement. Afterwards it should be ready to go.

Thanks for your patience, @erang20.

File: dts/bindings/dac/ti,dac-x0501.yaml (hunk 12 above)
Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Removed the default and made it required instead per @martinjaeger suggestion

File: dts/bindings/dac/ti,dac-x0501.yaml
[the remark this answers is not in the document]
Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml:1]: Done

File: dts/bindings/dac/ti,dac-x0501.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    description: |
+      DAC voltage reference select: either internal or external
+
+  voltage_mult:
+    type: string
+    enum:
+      - "mul2"
+      - "mul1"
+      - "div2"
+    default: "mul1"
```
[the remark this answers is not in the document]
Comment (erang20) [reply dts/bindings/dac/ti,dac-x0501.yaml]: Done

File: drivers/dac/dac_dacx0501.c [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+};
+
+struct dacx0501_config {
+	struct i2c_dt_spec i2c_spec;
+	enum ref_voltage_source ref_source;
+	enum voltage_mult voltage_mult;
+};
+
+struct dacx0501_data {
+	/* Device can only be configured once. */
```
[the remark this answers is not in the document]
Comment (erang20) [reply drivers/dac/dac_dacx0501.c:60]: Done

File: drivers/dac/dac_dacx0501.c (hunk 13 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c:148]: Done

File: drivers/dac/dac_dacx0501.c (hunk 14 above)
Comment (erang20) [reply drivers/dac/dac_dacx0501.c:110]: Done

File: drivers/dac/dac_dacx0501.c (hunk 13 above)
Comment (martinjaeger) [reply drivers/dac/dac_dacx0501.c:148]: Doesn't seem to be done.

File: dts/bindings/dac/ti,dacx0501.yaml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  output-gain:
+    type: string
+    required: true
+    enum:
+      - "mul2"
+      - "mul1"
+      - "div2"
+    description: |
+      This setting can be used to control the output voltage range within the supported bit
+      resolution. mul2 will double the output range but lower the resolutiuon, while div2 will
```
Comment (martinjaeger) [inline dts/bindings/dac/ti,dacx0501.yaml:29]: Small typo:

```suggestion
      resolution. mul2 will double the output range but lower the resolution, while div2 will
```

File: drivers/dac/dac_dacx0501.c [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+	if (status != 0) {
+		LOG_ERR("write CONFIG register failed");
+		return status;
+	}
+
+	status = dacx0501_reg_write(
+		dev, DACX0501_REG_GAIN,
+		FIELD_PREP(DACX0501_MASK_GAIN_REFDIV_EN, config->output_gain == VM_DIV2) |
+			FIELD_PREP(DACX0501_MASK_GAIN_BUFF_GAIN, config->output_gain == VM_MUL2));
+	if (status) {
```
Comment (martinjaeger) [inline drivers/dac/dac_dacx0501.c:164]:
```suggestion
	if (status != 0) {
```

**Review by martinjaeger — COMMENTED** [review]
Thanks for the quick updates. Only two very minor things left.

**Review by martinjaeger — APPROVED** [review]

**Review by bbilas — APPROVED** [review]

**Comment by erang20** [comment]
@martinjaeger any idea why merging is still blocked? Thanks :)

**Pull request merged**